### PR TITLE
exp push/pull: make all the flags in exp show and exp list support pull/push (#7154)

### PR DESCRIPTION
**Pull request by karajan1001**
wait for  #7245

exp push/pull: make all the flags in exp show and exp list support pull/push (#7154)

fix: #7154
1. add all flags (A/--all-commits , -a/--all-branches, --all-tags) in `exp show` to the `exp pull/push`
2. add unit and func tests for `exp push/pull`

* [x] ❗ I have followed the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) checklist.

* [x] 📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR (or issue, at least) in [dvc.org](https://github.com/iterative/dvc.org) and linked it here.

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

### CODE REVIEW

**Comment by karajan1001** [comment]
Demos for the push 
[asciicast](https://asciinema.org/a/34FZguA5R0XQa2iNQbwNsupUn)
for the pull
[asciicast](https://asciinema.org/a/gNcVfpqMTgnRxzvqg1GMznp62)

File: dvc/commands/experiments/utils.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
+def add_rev_selection_flags(
```
Comment (skshetry) [inline dvc/commands/experiments/utils.py:1]: You can move these to [`dvc/commands/experiments/__init__.py`](https://github.com/iterative/dvc/blob/47d10c4a8b9bdc18dec17a9380f0593e43383c4c/dvc/commands/experiments/__init__.py) and just import it as:
```python
from . import add_rev_selection_flags
```

**Comment by dberenbaum** [comment]
Nice work @karajan1001! I have a few questions:

1. Should there be an error if `dvc exp push/pull` are run without any arguments?

2. When there are no experiments to push, the output doesn't look right:
```
$ dvc exp push --no-cache myremote
Pushed experiment '[]'to Git remote 'myremote'.

To push cached outputs for this experiment to DVC remote storage,re-run this command without '--no-cache'.
```

We should recognize that no experiments have actually been pushed and make the output something like `No experiments pushed.`

3. `num` only seems to work when combined with `--rev`. Should it also be able to work from the current commit (like `dvc exp show -n 2`)?

File: dvc/commands/experiments/utils.py (hunk 1 above)
Comment (karajan1001) [reply dvc/commands/experiments/utils.py:1]: Yes but looks like that would cause a circular import, so we must place this import inside `add_parser`.

**Comment by karajan1001** [comment]
> 1. Should there be an error if `dvc exp push/pull` are run without any arguments?

In my opinion, For `exp push/pull/remove/gc` they are not a frequent cmd and will cause exp status changes, users must explicitly give the exps to be operated on, and should raise error if none arguments are given.

> 2. When there are no experiments to push, the output doesn't look right:
> 
> ```
> $ dvc exp push --no-cache myremote
> Pushed experiment '[]'to Git remote 'myremote'.
> 
> To push cached outputs for this experiment to DVC remote storage,re-run this command without '--no-cache'.
> ```
> 
> We should recognize that no experiments have actually been pushed and make the output something like `No experiments pushed.`

Yeah.

> 3. `num` only seems to work when combined with `--rev`. Should it also be able to work from the current commit (like `dvc exp show -n 2`)?

Yes, it works, `dvc exp show -n 2` will show the exps derive from `HEAD` and `HEAD~`

**Comment by karajan1001** [comment]
> > In my opinion, For `exp push/pull/remove/gc` they are not a frequent cmd and will cause exp status changes, users must explicitly give the exps to be operated on, and should raise error if none arguments are given.
> 
> 👍
> 
> > > We should recognize that no experiments have actually been pushed and make the output something like `No experiments pushed.`
> > 
> > 
> > Yeah.
> 
> 👍
> 
> > > 3. `num` only seems to work when combined with `--rev`. Should it also be able to work from the current commit (like `dvc exp show -n 2`)?
> > 
> > 
> > Yes, it works, `dvc exp show -n 2` will show the exps derive from `HEAD` and `HEAD~`
> 
> It works for me in `dvc exp show` but not in these other commands. For example:
> 
> ```
> $ dvc exp pull -n 2 myremote
> Pulled experiment '[]' from Git remote 'myremote'.
> $ dvc exp pull --rev HEAD -n 2 myremote
> Pulled experiment '['exp-45e9f']' from Git remote 'myremote'.
> ```

Are you in the latest version? I had already removed the default `HEAD` support. And this is the latest result from my computer.

```
> dvc exp pull -n 2 myremote
ERROR: Either of `experiment`, `--all-commits` or `--rev` needs to be set.
> dvc exp pull --rev HEAD -n 2 origin
Pulled experiment '['exp0.15', 'exp0.25', 'exp0.3', 'exp0.1']' from Git remote 'origin'.
>  dvc exp pull --rev HEAD~~ origin
No experiments to pull.
```

**Comment by dberenbaum** [comment]
Sorry, must have been behind. Looks good now! One last thing:

```
$ dvc exp pull
ERROR: the following arguments are required: <git_remote>, <experiment>
```

Should `<experiment>` be included as a required argument?

**Comment by karajan1001** [comment]
> Sorry, must have been behind. Looks good now! One last thing:
> 
> ```
> $ dvc exp pull
> ERROR: the following arguments are required: <git_remote>, <experiment>
> ```
> 
> Should `<experiment>` be included as a required argument?

This error comes from `argparse` I had set made `experiment` to be `*` which means it can be 0 to any integer.

**Comment by dberenbaum** [comment]
> add a default value for the arguments can solve this problem

Can the default be `None`?

**Comment by karajan1001** [comment]
> > add a default value for the arguments can solve this problem
> 
> Can the default be `None`?

I had tried several cases on my computer. It works fine for me.

File: dvc/commands/experiments/__init__.py [hunk 2]
```diff
@@ -58,3 +58,43 @@ def add_parser(subparsers, parent_parser):
...
+        "-n",
+        "--num",
+        type=int,
+        default=1,
+        dest="num",
+        metavar="<num>",
+        help=(
+            f"{command} experiments from the `--rev` baseline and from `num` "
+            "commits before it. Give a negative value to include all the "
+            "parent commits."
```
Comment (jorgeorpinel) [inline dvc/commands/experiments/__init__.py:98]: See https://github.com/iterative/dvc/pull/7245#pullrequestreview-902522092

File: dvc/commands/experiments/pull.py [hunk 3]
```diff
@@ -3,28 +3,45 @@
...
 
 
 class CmdExperimentsPull(CmdBase):
+    def raise_error_if_all_disabled(self):
+        if not any(
+            [self.args.experiment, self.args.all_commits, self.args.rev]
+        ):
+            raise InvalidArgumentError(
+                "Either of `experiment`, `--all-commits` "
+                "or `--rev` needs to be set."
```
Comment (jorgeorpinel) [inline dvc/commands/experiments/pull.py:19]: Same as https://github.com/iterative/dvc/pull/7275#pullrequestreview-902546358

Comment (jorgeorpinel) [reply dvc/commands/experiments/pull.py:19]: > Other comments in that PR also apply here.

File: dvc/commands/experiments/__init__.py (hunk 2 above)
Comment (jorgeorpinel) [reply dvc/commands/experiments/__init__.py:98]: > Other comments in that PR also apply here.

File: dvc/commands/experiments/pull.py (hunk 3 above)
Comment (jorgeorpinel) [reply dvc/commands/experiments/pull.py:19]: > Other comments in that PR also apply here.

Comment (karajan1001) [reply dvc/commands/experiments/pull.py]: Also updated in push.py

**Review by jorgeorpinel — APPROVED** [review]
UI strings LGTM 👍🏼

**Comment by karajan1001** [comment]
@pmrowla This one needs to be merged first before #7275.

**Review by pmrowla — APPROVED** [review]

**Pull request merged**